libclamav/udf.c

```diff
@@ -0,0 +1,838 @@
+
+    if (skipEmptyDescriptors(ctx, idxp, lastOffsetp)) {
+        goto done;
+    }
+
+    idx        = *idxp;
+    lastOffset = *lastOffsetp;
+
+    buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
+    if (NULL == buffer) {
+        goto done;
+    }
+    lastOffset = idx;
+
+    test = (ImplementationUseVolumeDescriptor *)buffer;
+    if (IMPLEMENTATION_USE_VOLUME_DESCRIPTOR != test->tag.tagId) {
+        goto done;
+    }
+
+    ret = test;
+    idx += VOLUME_DESCRIPTOR_SIZE;
+
+done:
+    *idxp        = idx;
+    *lastOffsetp = lastOffset;
+
+    return ret;
+}
+
+static LogicalVolumeDescriptor *getLogicalVolumeDescriptor(cli_ctx *ctx, size_t *idxp, size_t *lastOffsetp)
+{
+    uint8_t *buffer               = NULL;
+    LogicalVolumeDescriptor *ret  = NULL;
+    LogicalVolumeDescriptor *test = NULL;
+    size_t idx                    = *idxp;
+    size_t lastOffset             = *lastOffsetp;
+
+    if (skipEmptyDescriptors(ctx, idxp, lastOffsetp)) {
+        goto done;
+    }
+
+    idx        = *idxp;
+    lastOffset = *lastOffsetp;
+
+    buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
+    if (NULL == buffer) {
+        goto done;
+    }
+    lastOffset = idx;
+
+    test = (LogicalVolumeDescriptor *)buffer;
+    if (LOGICAL_VOLUME_DESCRIPTOR != test->tag.tagId) {
+        goto done;
+    }
+
+    idx += VOLUME_DESCRIPTOR_SIZE;
+    ret = test;
+
+done:
+    *idxp        = idx;
+    *lastOffsetp = lastOffset;
+
+    return ret;
+}
+
+static PartitionDescriptor *getPartitionDescriptor(cli_ctx *ctx, size_t *idxp, size_t *lastOffsetp)
+{
+    uint8_t *buffer           = NULL;
+    PartitionDescriptor *ret  = NULL;
+    PartitionDescriptor *test = NULL;
+    size_t idx                = *idxp;
+    size_t lastOffset         = *lastOffsetp;
+
+    if (skipEmptyDescriptors(ctx, idxp, lastOffsetp)) {
+        goto done;
+    }
+
+    idx        = *idxp;
+    lastOffset = *lastOffsetp;
+
+    buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
+    if (NULL == buffer) {
+        goto done;
+    }
+    lastOffset = idx;
+
+    test = (PartitionDescriptor *)buffer;
+    if (PARTITION_DESCRIPTOR != test->tag.tagId) {
+        goto done;
+    }
+
+    ret = test;
+    idx += VOLUME_DESCRIPTOR_SIZE;
+
+done:
+    *idxp        = idx;
+    *lastOffsetp = lastOffset;
+
+    return ret;
+}
+
+static UnallocatedSpaceDescriptor *getUnallocatedSpaceDescriptor(cli_ctx *ctx, size_t *idxp, size_t *lastOffsetp)
+{
+    uint8_t *buffer                  = NULL;
+    UnallocatedSpaceDescriptor *ret  = NULL;
+    UnallocatedSpaceDescriptor *test = NULL;
+    size_t idx                       = *idxp;
+    size_t lastOffset                = *lastOffsetp;
+
+    if (skipEmptyDescriptors(ctx, idxp, lastOffsetp)) {
+        goto done;
+    }
+
+    idx        = *idxp;
+    lastOffset = *lastOffsetp;
+
+    buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
```

Comment:
I'm a little concerned if this needs to be unneeded. Or maybe use `fmap_need_off_once()` instead